android/src/main/java/com/swmansion/gesturehandler/core/RotationGestureDetector.kt

```diff
@@ -85,50 +120,109 @@ class RotationGestureDetector(private val gestureListener: OnRotationGestureList
+      previousAngle = eventAngle
+      isPaused = false
+      localLog(".unpause(eventAngle: $eventAngle) fulfilled")
+    }
+  }
+
   private fun finish() {
     if (isInProgress) {
+      isPaused = false
       isInProgress = false
       gestureListener?.onRotationEnd(this)
     }
   }
 
   fun onTouchEvent(event: MotionEvent): Boolean {
+    fun getLogInfo(): String {
+      val firstPointerIndex = event.findPointerIndex(pointerIds[0])
+      val secondPointerIndex = event.findPointerIndex(pointerIds[1])
+
+      return """
+        pointerIds[0]: ${pointerIds[0]}, pointerIds[1]: ${pointerIds[1]}
+        firstPointerIndex: $firstPointerIndex, secondPointerIndex: $secondPointerIndex
+        isInProgress: $isInProgress, isPaused: $isPaused
+        action: ${event.action}, actionIndex: ${event.actionIndex} pointerCount: ${event.pointerCount}
+        rotation: $rotation, previousAngle: $previousAngle
+        event: $event
+      """.trimIndent()
+    }
+    localLog("onTouchEvent START | ${getLogInfo()}", true)
+
     when (event.actionMasked) {
       MotionEvent.ACTION_DOWN -> {
+        localLog("ActionDown START | ${getLogInfo()}")
         isInProgress = false
         pointerIds[0] = event.getPointerId(event.actionIndex)
         pointerIds[1] = MotionEvent.INVALID_POINTER_ID
+        localLog("ActionDown END | ${getLogInfo()}")
       }
-      MotionEvent.ACTION_POINTER_DOWN -> if (!isInProgress) {
-        pointerIds[1] = event.getPointerId(event.actionIndex)
-        isInProgress = true
-        previousTime = event.eventTime
-        previousAngle = Double.NaN
-        updateCurrent(event)
-        gestureListener?.onRotationBegin(this)
+      MotionEvent.ACTION_POINTER_DOWN -> {
+        localLog("ActionPointerDown START | ${getLogInfo()}")
+        if (!isInProgress) {
+          pointerIds[1] = event.getPointerId(event.actionIndex)
+          isInProgress = true
+          previousTime = event.eventTime
+          previousAngle = Double.NaN
+          updateCurrent(event)
+          gestureListener?.onRotationBegin(this)
+        }
+        localLog("ActionPointerDown END | ${getLogInfo()}")
       }
       MotionEvent.ACTION_MOVE -> if (isInProgress) {
-        updateCurrent(event)
-        gestureListener?.onRotation(this)
-      }
-      MotionEvent.ACTION_POINTER_UP -> {
-        val pointerId = event.getPointerId(event.actionIndex)
+        localLog("ActionMove START | ${getLogInfo()}", true)
+        val firstPointerIndex = event.findPointerIndex(pointerIds[0])
+        val secondPointerIndex = event.findPointerIndex(pointerIds[1])
 
-        // All key pointers are up
-        if (!isInProgress && pointerId == pointerIds[0]) {
-          gestureListener?.onRotationEnd(this)
+        if (
+          firstPointerIndex != MotionEvent.INVALID_POINTER_ID &&
+          secondPointerIndex != MotionEvent.INVALID_POINTER_ID
+        ) {
+          updateCurrent(event)
+          gestureListener?.onRotation(this)
         }
```

Comment:
`updateCurrent` already checks for invalid pointers, this way it's clearer what this part does

```suggestion
        updateCurrent(event)

        if (!isPaused) {
          gestureListener?.onRotation(this)
        }
```